src/dune_action_plugin/dune_action_plugin.ml

```diff
@@ -0,0 +1,203 @@
+      try Ok (f ()) with
+      | Unix.Unix_error (e, _, _) -> Error (name ^ ": " ^ Unix.error_message e)
+      | Sys_error error -> Error (name ^ ": " ^ error)
+
+    let read_directory =
+      let rec loop dh acc =
+        match Unix.readdir dh with
+        | "."
+        |".." ->
+          loop dh acc
+        | s -> loop dh (s :: acc)
+        | exception End_of_file -> acc
+      in
+      fun path ->
+        catch_system_exceptions ~name:"read_directory" (fun () ->
+          let dh = Unix.opendir path in
+          Stdune.Exn.protect
+            ~f:(fun () -> loop dh [] |> List.sort String.compare)
+            ~finally:(fun () -> Unix.closedir dh))
+
+    let read_file path =
+      catch_system_exceptions ~name:"read_file" (fun () ->
+        Stdune.Io.String_path.read_file path)
+
+    let write_file path data =
+      catch_system_exceptions ~name:"write_file" (fun () ->
+        Stdune.Io.String_path.write_file path data)
+  end
+
+  module Stage = struct
+    type 'a t =
+      { action : unit -> 'a
+      ; dependencies : Dependency.Set.t
+      ; targets : Stdune.String.Set.t
+      }
+
+    let map (t : 'a t) ~f = { t with action = (fun () -> f (t.action ())) }
+
+    let both (t1 : 'a t) (t2 : 'b t) =
+      { action = (fun () -> (t1.action (), t2.action ()))
+      ; dependencies = Dependency.Set.union t1.dependencies t2.dependencies
+      ; targets = Stdune.String.Set.union t1.targets t2.targets
+      }
+  end
+
+  (* Construction inspired by free monad. *)
+  type 'a t =
+    | Pure of 'a
+    | Stage of 'a t Stage.t
+
+  let lift_stage stage = Stage (Stage.map stage ~f:(fun a -> Pure a))
+
+  let rec map (t : 'a t) ~f =
+    match t with
+    | Pure a -> Pure (f a)
+    | Stage at -> Stage (Stage.map ~f:(map ~f) at)
+
+  let rec stage (t : 'a t) ~f =
+    match t with
+    | Pure a -> f a
+    | Stage at -> Stage (Stage.map ~f:(stage ~f) at)
+
+  let return a = Pure a
+
+  let rec both (t1 : 'a t) (t2 : 'b t) =
+    match (t1, t2) with
+    | Pure a1, _ -> map ~f:(fun a2 -> (a1, a2)) t2
+    | _, Pure a2 -> map ~f:(fun a1 -> (a1, a2)) t1
+    | Stage at1, Stage at2 ->
+      Stage (Stage.both at1 at2 |> Stage.map ~f:(fun (am1, am2) -> both am1 am2))
+
+  let read_file ~path =
+    let path = Path.to_string path in
+    let action () = Fs.read_file path |> Execution_error.raise_on_fs_error in
+    lift_stage
+      { action
+      ; dependencies = Dependency.Set.singleton (File path)
+      ; targets = Stdune.String.Set.empty
+      }
+
+  let write_file ~path ~data =
+    let path = Path.to_string path in
+    let action () =
+      Fs.write_file path data |> Execution_error.raise_on_fs_error
+    in
+    lift_stage
+      { action
+      ; dependencies = Dependency.Set.empty
+      ; targets = Stdune.String.Set.singleton path
+      }
+
+  (* TODO jstaron: If program tries to read empty directory, dune does not copy
+     it to `_build` so we get a "No such file or directory" error. *)
```

Comment:
It seems useful to document this behaviour in the API docs.

Reply:
Added a comment.